src/compiler/checker.ts

```diff
@@ -7276,6 +7276,82 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
+
+            // For regular function/method declarations, the enclosing declaration will already be signature.declaration,
+            // so this is a no-op, but for arrow functions and function expressions, the enclosing declaration will be
+            // the declaration that the arrow function / function expression is assigned to.
+            //
+            // If the parameters or return type include "typeof globalThis.paramName", using the wrong scope will lead
+            // us to believe that we can emit "typeof paramName" instead, even though that would refer to the parameter,
+            // not the global. Make sure we are in the right scope by changing the enclosingDeclaration to the function.
+            //
+            // We can't use the declaration directly; it may be in another file and so we may lose access to symbols
+            // accessible to the current enclosing declaration, or gain access to symbols not accessible to the current
+            // enclosing declaration. To keep this chain accurate, insert a fake scope into the chain which makes the
+            // function's parameters visible.
+            //
+            // If the declaration is in a JS file, then we don't need to do this at all, as there are no annotations besides
+            // JSDoc, which are always outside the function declaration, so are not in the parameter scope.
+            let cleanup: (() => void) | undefined;
+            if (
+                context.enclosingDeclaration
+                && signature.declaration
+                && signature.declaration !== context.enclosingDeclaration
+                && !isInJSFile(signature.declaration)
+                && some(expandedParams)
+            ) {
+                // As a performance optimization, share the symbol table between all faked scopes
+                // in this node builder chain. This is especially needed when we are working on an
+                // excessively deep type; if we don't do this, then we spend all of our time adding
+                // more and more scopes that need to be searched in isSymbolAccessible later. Since
+                // all we really want to do is to mark certain names as unavailable, we can just
+                // keep all of the names we're introducing in one large table and push/pop from it as
+                // needed; isSymbolAccessible will walk upward and find the closest "fake" scope,
+                // which will conveniently report on any and all faked scopes in the chain.
+                //
+                // It'd likely be better to store this somewhere else for isSymbolAccessible, but
+                // since that API _only_ uses the enclosing declaration (and its parents), this is
+                // seems like the best way to inject names into that search process.
+                const existingFakeScope = findAncestor(context.enclosingDeclaration, node => !!getNodeLinks(node).fakeScopeForSignatureDeclaration);

```

Comment:
I struggle to think of when you'd have another synthetic scope set as the child scope of an existing fake scope now. Maybe this can just be `!!getNodeLinks(context.enclosingDeclaration).fakeScopeForSignatureDeclaration` ?

Reply:
The case I'm worried about is the other places we reassign `context.enclosingDeclaration` temporarily, which happens in `serializeTypeAlias`, `sanitizeJSDocImplements`, `serializeAsClass`, `addPropertyToElementList`, etc.

I'd have to come up with an example of this, though.

Reply:
Okay, it does work to only ever check the most immediate enclosing declaration, at least in the the current tests. I would have thought it'd break, but examining things again, I can't actually figure out a counterexample here, so I'll switch things again.

Reply:
And to clarify, I did add an assert that verifies that there aren't more than 1 fake scope in a given chain, and it never fired, so that's proof enough I suppose.

Reply:
Switched to only check the current enclosingDeclaration.

Reply:
Just started reviewing this -- does this mean the comment block above also needs to be simplified?

Reply:
Yes, let me fix the wording. Most of it still applies but now there's only one fake scope.